gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java

```diff
@@ -132,29 +162,93 @@ public static void renamePath(FileSystem fs, Path oldName, Path newName) throws
    */
-  public static void copyPath(FileSystem fs, Path src, Path dst) throws IOException {
-    if (!FileUtil.copy(fs, src, fs, dst, false, fs.getConf())) {
-      throw new IOException(String.format("Failed to copy %s to %s", src, dst));
+  public static void copyPath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst, Configuration conf) throws IOException {
+
+    Preconditions.checkArgument(srcFs.exists(src),
+        String.format("Cannot copy from %s to %s because src does not exist", src, dst));
+    Preconditions
+        .checkArgument(!dstFs.exists(dst), String.format("Cannot copy from %s to %s because dst exists", src, dst));
+
+    try {
+      if (!FileUtil.copy(srcFs, src, dstFs, dst, false, false, conf)) {
+        throw new IOException(String.format("Failed to copy %s to %s", src, dst));
+      }
+    } catch (Exception e) {
+      try {
+        deleteIfExists(dstFs, dst, true);
+      } catch (IOException ioe) {
+        // Do nothing
+      }
+      throw e;
     }
   }
 
   /**
-   * A wrapper around {@link HadoopUtils#renamePath(FileSystem, Path, Path)} and
-   * {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} which will rename the path
-   * if the src and dst filesystems are the same; otherwise, the src file will be moved the the dst filesystem.
-   * An {@link IOException} if {@link FileUtil#copy(FileSystem, Path, FileSystem, Path, boolean, Configuration)} returns false.
+   * Copy a file from a srcFs {@link FileSystem} to a dstFs {@link FileSystem}. The src {@link Path} must be a file,
+   * that is {@link FileSystem#isFile(Path)} must return true for src.
+   *
+   * @param srcFs the src {@link FileSystem} to copy the file from
+   * @param src the src {@link Path} to copy
+   * @param dstFs the destination {@link FileSystem} to write to
+   * @param dst the destination {@link Path} to write to
+   * @param overwrite true if the dst {@link Path} should be overwritten, false otherwise
    */
-  public static void movePath(FileSystem srcFs, Path src, FileSystem dstFs, Path dst) throws IOException {
-    if (srcFs.getUri().equals(dstFs.getUri())) {
-      renamePath(srcFs, src, dst);
-    } else {
-      if (!FileUtil.copy(srcFs, src, dstFs, dst, true, false, dstFs.getConf())) {
-        throw new IOException(String.format("Failed to move %s to %s", src, dst));
+  public static void copyFile(FileSystem srcFs, Path src, FileSystem dstFs, Path dst, boolean overwrite,
+      Configuration conf) throws IOException {
+
+    Preconditions.checkArgument(srcFs.isFile(src),
+        String.format("Cannot copy from %s to %s because src is not a file", src, dst));
+    Preconditions.checkArgument(overwrite || !dstFs.exists(dst),
+        String.format("Cannot copy from %s to %s because dst exists", src, dst));
+
+    try (InputStream in = srcFs.open(src); OutputStream out = dstFs.create(dst, overwrite)) {
+      IOUtils.copyBytes(in, out, conf, false);
```

Comment:
This seems to have the same problems as before. If this line fails, `current.jst` will either be a partially copied file (if line 247 is not executed, e.g., SLA kill or OOM) or gone.

There doesn't seem to be an easy way to guarantee that if the new `current.jst` isn't successfully generated, the previous `current.jst` must be restored. So I still think modifying `FsDatasetStateStore.getLatestDatasetState` is a better approach. I don't think it violates any contract, because `current.jst` is just a convenient way to find the latest `jst` file (so that we don't need to list all `jst` files and sort them). If somehow `current.jst` failed to be generated, I don't see a problem of using the latest `jst` file.